examples/chef/common/clusters/channel/ChannelManager.cpp

```diff
@@ -0,0 +1,343 @@
+    program3.channel    = abc;
+    program3.title      = chip::CharSpan::fromCharString("My Program Title3");
+    program3.subtitle   = MakeOptional(chip::CharSpan::fromCharString("My Program Subtitle3"));
+    program3.startTime  = 0;
+    program3.endTime    = 60 * 60;
+
+    mPrograms.push_back(program3);
+}
+
+CHIP_ERROR ChannelManager::HandleGetChannelList(AttributeValueEncoder & aEncoder)
+{
+    // TODO: Insert code here
+    return aEncoder.EncodeList([](const auto & encoder) -> CHIP_ERROR {
+        for (auto const & channel : ChannelManager().mChannels)
+        {
+            ReturnErrorOnFailure(encoder.Encode(channel));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR ChannelManager::HandleGetLineup(AttributeValueEncoder & aEncoder)
+{
+    LineupInfoType lineup;
+    lineup.operatorName   = chip::CharSpan::fromCharString("Comcast");
+    lineup.lineupName     = MakeOptional(chip::CharSpan::fromCharString("Comcast King County"));
+    lineup.postalCode     = MakeOptional(chip::CharSpan::fromCharString("98052"));
+    lineup.lineupInfoType = chip::app::Clusters::Channel::LineupInfoTypeEnum::kMso;
+
+    return aEncoder.Encode(lineup);
+}
+
+CHIP_ERROR ChannelManager::HandleGetCurrentChannel(AttributeValueEncoder & aEncoder)
+{
+    return aEncoder.Encode(mCurrentChannel);
+}
+
+bool isChannelMatched(const ChannelInfoType & channel, const CharSpan & match)
+{
+    char number[32];
+    sprintf(number, "%d.%d", channel.majorNumber, channel.minorNumber);
+    bool nameMatch = channel.name.HasValue() ? channel.name.Value().data_equal(match) : false;
+    bool affiliateCallSignMatch =
+        channel.affiliateCallSign.HasValue() ? channel.affiliateCallSign.Value().data_equal(match) : false;
+    bool callSignMatch = channel.callSign.HasValue() ? channel.callSign.Value().data_equal(match) : false;
+    bool numberMatch   = match.data_equal(chip::CharSpan::fromCharString(number));
+
+    return affiliateCallSignMatch || callSignMatch || nameMatch || numberMatch;
+}
+
+void ChannelManager::HandleChangeChannel(CommandResponseHelper<ChangeChannelResponseType> & helper, const CharSpan & match)
+{
+    std::vector<ChannelInfoType> matchedChannels;
+    uint16_t index = 0;
+    for (auto const & channel : mChannels)
+    {
+        // verify if CharSpan matches channel name
+        // or callSign or affiliateCallSign or majorNumber.minorNumber
+        if (isChannelMatched(channel, match))
+        {
+            matchedChannels.push_back(channel);
+        }
+        else if (matchedChannels.size() == 0)
+        {
+            // "index" is only used when we end up with matchedChannels.size() == 1.
+            // In that case, we want it to be the number of non-matching channels we saw before
+            // the matching one.
+            index++;
+        }
+    }
+
+    ChangeChannelResponseType response;
+
+    // Error: Found multiple matches
+    if (matchedChannels.size() > 1)
```

Comment:
why do we go through all channels to find this? If this is a critical error it could be done during the loop: if we match and we already had a match (vector is non-empty) then return an error.

Further more, we do not need a vector anymore at all - we only need to store one match.

Reply:
matchedChannels removed.